libtransmission/torrent.cc

```diff
@@ -1775,10 +1775,16 @@ void tr_torrent::recheck_completeness()
         }
 
         completeness_ = new_completeness;
-        session->close_torrent_files(id());
+
+        if (is_done() && was_running && session->shouldFullyVerifyCompleteTorrents())
```

Comment:
No, I thought this through and tested it previously.
We should verify before closing the torrent, in order to keep the files opened in case it wasn't complete.
And `is_done()` needs to be evaluated a second time after `tr_torrentVerify`.

Regarding `recent_change`, we should also not rely on it. It can only be false if running completion happened with zero bytes downloaded in the session, which can only happen if we _unselected_ files to be downloaded, which is still a scenario where we do want to verify the data.

So there are no changes needed to my proposed code.

Reply:
I wrote up a response to this, complete with code suggestion, but later realised it may not be relevant... See https://github.com/transmission/transmission/pull/4178#issuecomment-2071463065

In any case, I think it makes sense to move `session->close_torrent_files(id())` inside `if (is_done())`.

<details>
<summary>Old (outdated) response</summary>

I can see where Charles is coming from, the repeated if conditions bothers me a bit as well.

I propose something like

```c++
if (is_done())
{
    session->close_torrent_files(id());

    if (was_running && session->shouldFullyVerifyCompleteTorrents())
    {
        tr_torrentVerify(this);
    }

    if (recent_change)
    {

    // ...
```

> We should verify before closing the torrent, in order to keep the files opened in case it wasn't complete.

I don't think this matters. Calling `tr_torrentVerify()` calls `tr_torrent::stop_now()`, which calls `tr_session::close_torrent_files()` anyway.

> And `is_done()` needs to be evaluated a second time after `tr_torrentVerify`

I cannot see why. `tr_torrentVerify()` is non-blocking, so checking `is_done()` after the call isn't very useful. Also `tr_torent::recheck_completeness()` will be called again after the verify completes.

> Regarding `recent_change`, we should also not rely on it. It can only be false if running completion happened with zero bytes downloaded in the session, which can only happen if we _unselected_ files to be downloaded, which is still a scenario where we do want to verify the data.

This I agree, it checks out.

</details>

Reply:
I understand now. Fix at https://github.com/transmission/transmission/commit/73a20c6cdde531feaead283dd2cbd8b869044b3b